File: app/src/main/java/me/imzack/app/end/view/fragment/MyPlansFragment.java
package me.imzack.app.end.view.fragment;

import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import me.imzack.app.end.App;

import me.imzack.app.end.R;
import me.imzack.app.end.injector.component.DaggerMyPlansComponent;
import me.imzack.app.end.injector.module.MyPlansPresenterModule;
import me.imzack.app.end.view.activity.PlanDetailActivity;
import me.imzack.app.end.view.adapter.PlanListAdapter;
import me.imzack.app.end.view.contract.MyPlansViewContract;
import me.imzack.app.end.presenter.MyPlansPresenter;
import me.imzack.app.end.model.bean.Plan;

import javax.inject.Inject;

import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;

public class MyPlansFragment extends BaseListFragment implements MyPlansViewContract {

    @BindView(R.id.list_plan)
    RecyclerView mPlanList;
    @BindView(R.id.layout_empty)
    LinearLayout mEmptyLayout;

    @BindString(R.string.snackbar_delete_format)
    String mSnackbarDeleteFormat;

    @Inject
    MyPlansPresenter mMyPlansPresenter;

    public MyPlansFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onInjectPresenter() {
        DaggerMyPlansComponent.builder()
                .myPlansPresenterModule(new MyPlansPresenterModule(this))
                .appComponent(App.getAppComponent())
                .build()
                .inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_my_plans, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);
        mMyPlansPresenter.attach();
    }

    @Override
    public void onResume() {
        super.onResume();
        mMyPlansPresenter.notifySwitchingViewVisibility(true);
    }

    @Override
    public void onPause() {
        super.onPause();
        mMyPlansPresenter.notifySwitchingViewVisibility(false);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mMyPlansPresenter.detach();
    }

    @Override
    public void showInitialView(PlanListAdapter planListAdapter, ItemTouchHelper itemTouchHelper, boolean isPlanItemEmpty) {
        mPlanList.setAdapter(planListAdapter);
        mPlanList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                onListScrolled(dy);
                mMyPlansPresenter.notifyPlanListScrolled(
                        !mPlanList.canScrollVertically(-1),
                        !mPlanList.canScrollVertically(1)
                );
            }
        });
        itemTouchHelper.attachToRecyclerView(mPlanList);

        onPlanItemEmptyStateChanged(isPlanItemEmpty);
    }

    @Override
    public void onPlanItemClicked(int position) {
        PlanDetailActivity.start(getContext(), position);
    }

    @Override
    public void onPlanCreated() {
        mPlanList.scrollToPosition(0);
    }

    @Override
    public void onPlanDeleted(final Plan deletedPlan, final int position, boolean shouldShowSnackbar) {
        if (shouldShowSnackbar) {
            Snackbar.make(mPlanList, String.format(mSnackbarDeleteFormat, deletedPlan.getContent()), Snackbar.LENGTH_LONG)
                    .setAction(R.string.button_undo, new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            mMyPlansPresenter.notifyCreatingPlan(deletedPlan, position);
                        }
                    })
                    .show();
        }
    }

    @Override
    public void onPlanItemEmptyStateChanged(boolean isEmpty) {
        mPlanList.setVisibility(isEmpty ? View.GONE : View.VISIBLE);
        mEmptyLayout.setVisibility(isEmpty ? View.VISIBLE : View.GONE);
    }

    @Override
    public void exit() {
        remove();
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/fragment/MyPlansFragment.kt
package net.zackzhang.app.end.view.fragment

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_my_plans.*
import net.zackzhang.app.end.App
import net.zackzhang.app.end.R
import net.zackzhang.app.end.injector.component.DaggerMyPlansComponent
import net.zackzhang.app.end.injector.module.MyPlansPresenterModule
import net.zackzhang.app.end.model.bean.Plan
import net.zackzhang.app.end.presenter.MyPlansPresenter
import net.zackzhang.app.end.view.activity.PlanDetailActivity
import net.zackzhang.app.end.view.adapter.PlanListAdapter
import net.zackzhang.app.end.view.contract.MyPlansViewContract
import javax.inject.Inject

class MyPlansFragment : BaseListFragment(), MyPlansViewContract {

    @Inject
    lateinit var mMyPlansPresenter: MyPlansPresenter

    override fun onInjectPresenter() {
        DaggerMyPlansComponent.builder()
                .myPlansPresenterModule(MyPlansPresenterModule(this))
                .appComponent(App.appComponent)
                .build()
                .inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            inflater.inflate(R.layout.fragment_my_plans, container, false)!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mMyPlansPresenter.attach()
    }

    override fun onResume() {
        super.onResume()
        mMyPlansPresenter.notifySwitchingViewVisibility(true)
    }

    override fun onPause() {
        super.onPause()
        mMyPlansPresenter.notifySwitchingViewVisibility(false)
    }

    override fun onDetach() {
        super.onDetach()
        mMyPlansPresenter.detach()
    }

    override fun showInitialView(planListAdapter: PlanListAdapter, itemTouchHelper: ItemTouchHelper, isPlanItemEmpty: Boolean) {
        list_plan.adapter = planListAdapter
        list_plan.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                onListScrolled(dy)
                mMyPlansPresenter.notifyPlanListScrolled(
                        !list_plan.canScrollVertically(-1),
                        !list_plan.canScrollVertically(1)
                )
            }
        })
        itemTouchHelper.attachToRecyclerView(list_plan)

        onPlanItemEmptyStateChanged(isPlanItemEmpty)
    }

    override fun onPlanItemClicked(position: Int) {
        PlanDetailActivity.start(context!!, position)
    }

    override fun onPlanCreated() {
        list_plan.scrollToPosition(0)
    }

    override fun onPlanDeleted(deletedPlan: Plan, position: Int, shouldShowSnackbar: Boolean) {
        if (shouldShowSnackbar) {
            Snackbar.make(list_plan, String.format(getString(R.string.snackbar_delete_format), deletedPlan.content), Snackbar.LENGTH_LONG)
                    .setAction(R.string.button_undo) { mMyPlansPresenter.notifyCreatingPlan(deletedPlan, position) }
                    .show()
        }
    }

    override fun onPlanItemEmptyStateChanged(isEmpty: Boolean) {
        list_plan.visibility = if (isEmpty) View.GONE else View.VISIBLE
        layout_empty.visibility = if (isEmpty) View.VISIBLE else View.GONE
    }
}
